File: omnivote/omnivote/lib/services/blockchain_service.dart
import 'dart:convert';
import 'dart:math';
import 'package:crypto/crypto.dart';
import '../models/models.dart';
import '../constants/app_constants.dart';

// Placeholder service for blockchain operations
// In production, this would integrate with actual Solana blockchain
class BlockchainService {
  // Simulate generating a DID (Decentralized Identity)
  Future<String> generateDID() async {
    // Simulate async operation
    await Future.delayed(const Duration(milliseconds: 500));
    
    // Generate a mock DID
    final random = Random.secure();
    final bytes = List<int>.generate(32, (i) => random.nextInt(256));
    final hash = sha256.convert(bytes);
    
    return 'did:omnivote:${hash.toString().substring(0, 42)}';
  }
  
  // Generate public/private key pair
  Future<Map<String, String>> generateKeyPair() async {
    await Future.delayed(const Duration(milliseconds: 500));
    
    // Mock key generation (in production, use actual crypto libraries)
    final random = Random.secure();
    final publicKey = List<int>.generate(32, (i) => random.nextInt(256));
    final privateKey = List<int>.generate(64, (i) => random.nextInt(256));
    
    return {
      'publicKey': base64Encode(publicKey),
      'privateKey': base64Encode(privateKey),
    };
  }
  
  // Cast a vote on the blockchain
  Future<String> castVote({
    required String electionId,
    required String candidateId,
    required String voterId,
    String? zkProof,
  }) async {
    // Simulate transaction time
    await Future.delayed(const Duration(seconds: 2));
    
    // Generate mock transaction hash
    final timestamp = DateTime.now().millisecondsSinceEpoch;
    final data = '$electionId$candidateId$voterId$timestamp';
    final hash = sha256.convert(utf8.encode(data));
    
    return hash.toString();
  }
  
  // Verify a vote transaction
  Future<VoteStatus> verifyTransaction(String transactionHash) async {
    await Future.delayed(const Duration(milliseconds: 800));
    
    // Mock verification - in production, check actual blockchain
    return VoteStatus.confirmed;
  }
  
  // Get transaction confirmation blocks
  Future<int> getConfirmationBlocks(String transactionHash) async {
    await Future.delayed(const Duration(milliseconds: 500));
    
    // Mock confirmation blocks
    return Random().nextInt(10) + 1;
  }
  
  // Generate zero-knowledge proof
  Future<String> generateZKProof({
    required String voterId,
    required String electionId,
  }) async {
    await Future.delayed(const Duration(milliseconds: 1000));
    
    // Mock ZK proof generation
    final data = '$voterId$electionId${DateTime.now().millisecondsSinceEpoch}';
    final hash = sha256.convert(utf8.encode(data));
    
    return 'zkp:${hash.toString().substring(0, 64)}';
  }
  
  // Verify eligibility without revealing identity
  Future<bool> verifyEligibility({
    required String voterId,
    required String electionId,
    required String zkProof,
  }) async {
    await Future.delayed(const Duration(milliseconds: 800));
    
    // Mock eligibility verification
    return true;
  }
  
  // Get transaction details
  Future<Map<String, dynamic>> getTransactionDetails(
    String transactionHash,
  ) async {
    await Future.delayed(const Duration(milliseconds: 600));
    
    return {
      'hash': transactionHash,
      'timestamp': DateTime.now().toIso8601String(),
      'status': 'confirmed',
      'blockNumber': Random().nextInt(1000000) + 1000000,
      'confirmations': Random().nextInt(10) + 3,
      'fee': '0.00025',
      'network': 'Solana Mainnet',
    };
  }
  
  // Estimate transaction fee
  Future<double> estimateTransactionFee() async {
    await Future.delayed(const Duration(milliseconds: 300));
    
    // Return Solana's typical transaction fee
    return 0.00025;
  }
  
  // Check if wallet has sufficient balance
  Future<bool> checkBalance(String publicKey, double requiredAmount) async {
    await Future.delayed(const Duration(milliseconds: 400));
    
    // Mock balance check
    return true;
  }
  
  // Get wallet balance
  Future<double> getBalance(String publicKey) async {
    await Future.delayed(const Duration(milliseconds: 400));
    
    // Mock balance (in SOL)
    return 1.5 + Random().nextDouble() * 0.5;
  }
}

// Service for managing user's blockchain wallet
class WalletService {
  String? _publicKey;
  String? _privateKey;
  
  Future<void> createWallet() async {
    final keyPair = await BlockchainService().generateKeyPair();
    _publicKey = keyPair['publicKey'];
    _privateKey = keyPair['privateKey'];
  }
  
  String? get publicKey => _publicKey;
  
  Future<void> loadWallet(String publicKey, String privateKey) async {
    _publicKey = publicKey;
    _privateKey = privateKey;
  }
  
  bool get hasWallet => _publicKey != null && _privateKey != null;
  
  void clearWallet() {
    _publicKey = null;
    _privateKey = null;
  }
}

File: omnivote-backend/src/services/blockchainService.js
const crypto = require('crypto');

class BlockchainService {
  /**
   * Generate a Decentralized Identity (DID)
   */
  async generateDID() {
    const bytes = crypto.randomBytes(32);
    const hash = crypto.createHash('sha256').update(bytes).digest('hex');
    return `did:omnivote:${hash.substring(0, 42)}`;
  }

  /**
   * Generate public/private key pair
   * FIX: Node.js crypto.generateKeyPairSync only accepts format: 'pem' or 'der'.
   *      'base64' is not a valid format — it was causing ERR_INVALID_ARG_VALUE.
   *      We now export as 'pem' and strip the PEM headers to get a clean base64 string.
   */
  async generateKeyPair() {
    const { publicKey, privateKey } = crypto.generateKeyPairSync('ec', {
      namedCurve: 'secp256k1',
      publicKeyEncoding: { type: 'spki', format: 'pem' },
      privateKeyEncoding: { type: 'pkcs8', format: 'pem' }
    });

    // Strip PEM headers/footers and newlines to get raw base64
    const toBase64 = (pem) => pem
      .replace(/-----BEGIN [^-]+-----/, '')
      .replace(/-----END [^-]+-----/, '')
      .replace(/\s+/g, '');

    return {
      publicKey: toBase64(publicKey),
      privateKey: toBase64(privateKey)
    };
  }

  /**
   * Cast a vote and return a transaction hash
   */
  async castVote({ electionId, candidateId, voterId, zkProof }) {
    // Simulate blockchain transaction delay
    await this._simulateDelay(1500);

    const data = `${electionId}${candidateId}${voterId}${Date.now()}${Math.random()}`;
    const transactionHash = crypto.createHash('sha256').update(data).digest('hex');

    return {
      transactionHash,
      blockNumber: Math.floor(Math.random() * 1000000) + 200000000,
      fee: 0.00025,
      network: 'Solana Mainnet',
      status: 'confirmed'
    };
  }

  /**
   * Verify a transaction on-chain
   */
  async verifyTransaction(transactionHash) {
    await this._simulateDelay(800);
    return {
      status: 'confirmed',
      confirmationBlocks: Math.floor(Math.random() * 7) + 3,
      blockNumber: Math.floor(Math.random() * 1000000) + 200000000,
      timestamp: new Date().toISOString()
    };
  }

  /**
   * Generate zero-knowledge proof
   */
  async generateZKProof({ voterId, electionId }) {
    await this._simulateDelay(1000);
    const data = `${voterId}${electionId}${Date.now()}`;
    const proof = crypto.createHash('sha256').update(data).digest('hex');
    return `zkp:${proof}`;
  }

  /**
   * Verify eligibility using ZK proof
   */
  async verifyEligibility({ voterId, electionId, zkProof }) {
    await this._simulateDelay(600);
    // In production, verify against smart contract
    return { eligible: true, reason: null };
  }

  /**
   * Hash a voter's DID for anonymous storage
   */
  hashVoterDID(did) {
    return crypto.createHash('sha256').update(did + process.env.JWT_SECRET).digest('hex');
  }

  /**
   * Get estimated transaction fee
   */
  async estimateTransactionFee() {
    return 0.00025; // SOL
  }

  /**
   * Get wallet balance
   */
  async getBalance(publicKey) {
    await this._simulateDelay(400);
    return 1.5 + Math.random() * 0.5;
  }

  /**
   * Get full transaction details
   */
  async getTransactionDetails(transactionHash) {
    await this._simulateDelay(600);
    return {
      hash: transactionHash,
      timestamp: new Date().toISOString(),
      status: 'confirmed',
      blockNumber: Math.floor(Math.random() * 1000000) + 200000000,
      confirmations: Math.floor(Math.random() * 10) + 3,
      fee: '0.00025',
      network: 'Solana Mainnet'
    };
  }

  _simulateDelay(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}

module.exports = new BlockchainService();
